ocaml/testsuite/tests/typing-modes/stack.ml

```diff
@@ -0,0 +1,103 @@
+(* TEST
```

Comment:
I would love to see some more tests. Here are some ideas:

* Each of these cases where `stack_` is rejected should have a similar case nearby where `stack_` is accepted. Otherwise, we'll never know if `stack_` starts rejecting more things.
* There should be cases testing that `stack_` works deeply. That is, my understanding of the specification of `stack_` is that e.g. `stack_ (K1 (K2 (K3 x)))` is meant to work deeply and should mean that all of `K1`, `K2`, and `K3` are allocated on the stack.... which might not be true if one of them has a global modality. So I would want to see tests both with and without intervening global modalities.
* Is there a way to assert `stack_`-allocation on patterns that allocate? Probably not, and that's probably OK for now.
* I would want to see some tests on lists, because list literals actually encompass many allocations.
* Tests on polymorphic-variant constructors?

As for the lack of reasons/hints: I think this feature would be better with them. Stack-allocation is an advanced feature, and helping users learn it is a good thing. I don't imagine it would be too hard to add, along the lines you suggest in the PR description.

Reply:
> There should be cases testing that stack_ works deeply.

`stack_` works shallowly. For example:
```
let _ = ref x in
stack_ (K x)
```
will be accepted. `x` is on heap and `K x` is on stack.

Reply:
> Is there a way to assert stack_-allocation on patterns that allocate?

Not in this PR; but I don't see obvious difficulty. But I think we want to support modes on patterns first.

Reply:
> Tests on polymorphic-variant constructors?

I'm curious about your interest in this - I thought they are similar to regular constructors so I skipped.

Reply:
> I would want to see some tests on lists, because list literals actually encompass many allocations.

This wouldn't be super interesting for the reason I listed above - `stack_` is not deep.

Reply:
My understanding is that `stack_` would be deep on nested constructor calls. That is, `stack_ (Some (Some thing))` would ensure that both `Some`s are stack-allocated. Contrast with `stack_ G (G thing)` where we have `type ‘a t = G of ‘a @@ global`; the inner `G` would not be stack allocated.

But maybe my understanding is wrong? Might be good to raise this design issue on the appropriate Slack channel, because we want to agree on what would be most useful to users.

Reply:
My only interest in polymorphic variants is that there is a specific line of code to make `stack_` work on polymorphic variants, so I thought it would probably be good to test that it works.

Reply:
Also, we shouldn’t forget about tests that do not produce error messages.

Reply:
Yeah I can intuitively see how deep `stack_` can be useful. But in your second example, the `stack_` interact with modality, which I don't like. Because `stack_` operates on the axis `stack/heap` (which currently collides with `local/global`), while the `global` modality operates on the regionality axis.

Instead, we say: `stack_` works deeply no matter what. That means the user has to write `let x = G thing in stack_ (G x)`.

But this also means allowing user to write `stack_ (foo bar)` that doesn't do anything. Maybe we should require "each stack_ should at least do something, or we issue a warning"?

If the above doesn't deter you, let's bring this up to a wider audience.

Reply:
I thought it would be deep only over nested constructor calls, not passing inside of function application. To be honest, I don’t have a strong sense for what would be best here — I’m just reporting on what I thought was the agreed plan (but I don’t have a reference to where we made that agreement; it’s just what I remember).

And I realize, upon rereading, that my previous post about the global modality was confusing. I mean that `stack_ (G (G thing))` would report an error, while `stack_ (Some (Some thing))` would not report an error. This allows users to witness the global modality (which they may have forgotten).

Maybe the best thing would be for this patch to include a change to the documentation which describes how the feature works. Then we’ll have a specification of what it means and we can then seek wider opinions on whether that specification is the best one.

Reply:
good point, will do.

Reply:
I think I'm gonna skip changing the document, because currently the local allocation document strongly associates `local_` with stack allocation. Adding a section about `stack_` without changing the rest is not coherent.

I suggest that we merge this PR without telling the user. We will find time to overhaul the local allocation document, and then we can announce `stack_` to the user.

Reply:
Yet I still don't know that I crisply understand the specification of this feature, if there is no documentation. The tests can get at this specification, but no test really explores the "deep" aspect of this that I've been asking about in this thread. For example:

```
type t = { x : int list; y : int list @@ global }
let mk () = exclave_ stack_ { x = [1;2;3]; y = [4;5;6] }
```

I don't know whether this should be accepted or not.

Reply:
That's fair. I will add some tests demonstrating this aspect.

Reply:
Your example will be shadowed by much simpler tests, tho. Because `stack_` works shallowly, so the modalities on fields doesn't matter.